arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -0,0 +1,334 @@
+            typeWidth = UNKNOWN_WIDTH;
+            break;
+          case INT_8:
+          case INT_16:
+          case INT_32:
+            this.vec = ArrowSchemaUtil.convert(icebergField).createVector(rootAlloc);
+            ((IntVector) vec).allocateNew(batchSize);
+            readType = ReadType.INT;
+            typeWidth = IntVector.TYPE_WIDTH;
+            break;
+          case DATE:
+            this.vec = ArrowSchemaUtil.convert(icebergField).createVector(rootAlloc);
+            ((DateDayVector) vec).allocateNew(batchSize);
+            readType = ReadType.INT;
+            typeWidth = IntVector.TYPE_WIDTH;
+            break;
+          case INT_64:
+          case TIMESTAMP_MILLIS:
+            this.vec = ArrowSchemaUtil.convert(icebergField).createVector(rootAlloc);
+            ((BigIntVector) vec).allocateNew(batchSize);
+            readType = ReadType.LONG;
+            typeWidth = BigIntVector.TYPE_WIDTH;
+            break;
+          case TIMESTAMP_MICROS:
+            this.vec = ArrowSchemaUtil.convert(icebergField).createVector(rootAlloc);
+            ((TimeStampMicroTZVector) vec).allocateNew(batchSize);
+            readType = ReadType.LONG;
+            typeWidth = BigIntVector.TYPE_WIDTH;
+            break;
+          case DECIMAL:
+            DecimalMetadata decimal = primitive.getDecimalMetadata();
+            this.vec = new IcebergArrowVectors.DecimalArrowVector(icebergField.name(), rootAlloc,
+                decimal.getPrecision(), decimal.getScale());
+            ((DecimalVector) vec).allocateNew(batchSize);
+            switch (primitive.getPrimitiveTypeName()) {
+              case BINARY:
+              case FIXED_LEN_BYTE_ARRAY:
+                readType = ReadType.FIXED_LENGTH_DECIMAL;
+                typeWidth = primitive.getTypeLength();
+                break;
+              case INT64:
+                readType = ReadType.INT_LONG_BACKED_DECIMAL;
+                typeWidth = BigIntVector.TYPE_WIDTH;
+                break;
+              case INT32:
+                readType = ReadType.INT_LONG_BACKED_DECIMAL;
+                typeWidth = IntVector.TYPE_WIDTH;
+                break;
+              default:
+                throw new UnsupportedOperationException(
+                    "Unsupported base type for decimal: " + primitive.getPrimitiveTypeName());
+            }
+            break;
+          default:
+            throw new UnsupportedOperationException(
+                "Unsupported logical type: " + primitive.getOriginalType());
+        }
+      } else {
+        switch (primitive.getPrimitiveTypeName()) {
+          case FIXED_LEN_BYTE_ARRAY:
+            int len = ((Types.FixedType) icebergField.type()).length();
+            this.vec = new IcebergArrowVectors.VarBinaryArrowVector(icebergField.name(), rootAlloc);
+            int factor = (len + DEFAULT_RECORD_BYTE_COUNT - 1) / DEFAULT_RECORD_BYTE_COUNT;
+            vec.setInitialCapacity(batchSize * factor);
```

Comment:
It isn't clear why this uses factor instead of `len` to determine the initial allocation size. I think it should be number of records per batch * bytes per value (len). That means we can also get rid of `DEFAULT_RECORD_BYTE_COUNT` because it isn't used anywhere else.